### PR TITLE
Reduce UInt256 in EVMOperations

### PR DESCRIPTION
**Pull request by shemnon**
## PR description

Improve the performance of the EVM by reducing the usage of UInt256 in almost all operations and moving the main switch to a Java 17 expression switch

### CODE REVIEW

**Comment by shemnon** [comment]
Benchmark results - https://docs.google.com/spreadsheets/d/1qbvlAveoLCiWbhbxouHgcpFeDVjws7UD0dgZghJZ6oU/edit?usp=sharing

Macbook Pro M1 MAX

Top line result - 10-15% "all in"
Target opcodes - 100-500% improvement.  Most notably in JumpDest& SDIV

File: evm/src/main/java/org/hyperledger/besu/evm/operation/SDivOperation.java
Comment (macfarla) [inline evm/src/main/java/org/hyperledger/besu/evm/operation/SDivOperation.java:78]: did you mean to leave this commented code?

**Review by macfarla — COMMENTED** [review]
a few comments on parts I don't understand 
overall I think it's fine
performance improvement sounds fab 
think a second opinion would be good :)

**Comment by macfarla** [comment]
@shemnon should this get a changelog entry?

File: evm/src/test/java/org/hyperledger/besu/evm/operations/SarOperationTest.java
Comment (shemnon) [reply evm/src/test/java/org/hyperledger/besu/evm/operations/SarOperationTest.java:213]: That's something worth benchmarking, but the number of times we need to take a 256 bit number and consider it as a 8 or 32 bit number are very high (memory access, shifts, etc).  So my first pass was trimmed bytes, where we can assess size much quicker.

I'll look into Bytes32 or possibly ByteBuffer integrations in a future pass. It did cross my mind that consistently sized words may improve garbage collection implications and (by extension) speed.

File: evm/src/main/java/org/hyperledger/besu/evm/operation/ByteOperation.java
Comment (shemnon) [reply evm/src/main/java/org/hyperledger/besu/evm/operation/ByteOperation.java:41]: Semantically they are the same on the stack, it's quicker to get an int or byte zero out of a Bytes.EMPTY than a UInt256.ZERO (UInt256 would have to check 7 ints as zero before returning the last one, and possibly bit checking for size).

The tests that detected the Tuweni 3.2 constant values bug (where Bytes32.ZERO couldn't be trimmed to 20 bytes in length) would trip this if it was a problem.

File: evm/src/main/java/org/hyperledger/besu/evm/operation/SDivOperation.java
Comment (shemnon) [reply evm/src/main/java/org/hyperledger/besu/evm/operation/SDivOperation.java]: No, somehow it passed my sonar checks.  Will remove

File: evm/src/main/java/org/hyperledger/besu/evm/operation/AbstractCallOperation.java
Comment (garyschulte) [inline evm/src/main/java/org/hyperledger/besu/evm/operation/AbstractCallOperation.java:186]: FAILURE_STACK_ITEM here?

File: evm/src/main/java/org/hyperledger/besu/evm/operation/BlockHashOperation.java
Comment (garyschulte) [inline evm/src/main/java/org/hyperledger/besu/evm/operation/BlockHashOperation.java:46]: might this lead to problems if we are not aligned on a word boundary?

File: evm/src/main/java/org/hyperledger/besu/evm/operation/ByteOperation.java
Comment (macfarla) [reply evm/src/main/java/org/hyperledger/besu/evm/operation/ByteOperation.java:46]: ahhh I skimmed over the trimLeadingZeros. makes sense now.

File: evm/src/main/java/org/hyperledger/besu/evm/operation/JumpDestOperation.java
Comment (garyschulte) [inline evm/src/main/java/org/hyperledger/besu/evm/operation/JumpDestOperation.java:25]: suggest all caps on public static final

**Review by macfarla — APPROVED** [review]

**Review by garyschulte — APPROVED** [review]
the slicing is a bit scary tbh, but happy to see evm perf improvements.  This PR would be a good candidate for fuzzing on a private testnet.  What kind of testing did you do already?

File: evm/src/main/java/org/hyperledger/besu/evm/operation/BlockHashOperation.java
Comment (shemnon) [reply evm/src/main/java/org/hyperledger/besu/evm/operation/BlockHashOperation.java:46]: It won't.  The call to `blockArg.toLong()` on line 54 handles any byte length of 8 or less. That's the only use of this variable.

File: evm/src/main/java/org/hyperledger/besu/evm/operation/AbstractCallOperation.java
Comment (shemnon) [reply evm/src/main/java/org/hyperledger/besu/evm/operation/AbstractCallOperation.java:186]: done

File: evm/src/main/java/org/hyperledger/besu/evm/operation/JumpDestOperation.java
Comment (shemnon) [reply evm/src/main/java/org/hyperledger/besu/evm/operation/JumpDestOperation.java:25]: done

File: evm/src/main/java/org/hyperledger/besu/evm/operation/SubOperation.java
Comment (shemnon) [reply evm/src/main/java/org/hyperledger/besu/evm/operation/SubOperation.java:57]: I expect there are some JVM optimizations targeting that stretch of code.  It doesn't look like it's directly an @Intrinsic function but it ought to be.  Also, the zero check in UInt256 is kind of expensive.

**Comment by shemnon** [comment]
Before I press Merge I was going to give it a day or so on Marius's fuzzer.  It found the other corner cases fairly quickly.

**Comment by non-fungible-nelson** [comment]
How did it go? @shemnon

**Comment by shemnon** [comment]
Fuzz testing was useful.

**Pull request merged**
